File: RMG.Core/Music/Pattern.cs
using System.Collections.Generic;

namespace RMG.Core.Music
{
    public sealed class Pattern : IDuration
    {
        public IReadOnlyList<TimedEvent<Note>> Notes { get; set; }

        public IReadOnlyList<TimedEvent<Pattern>> Patterns { get; set; }

        public NoteBasePattern NoteBasePattern { get; set; }

        public double Duration { get; set; }
    }
}

File: rmg-js/src/composition/pattern.ts
import { Timeline } from '../core/timeline';
import { DurationEntity } from '../music/duration-entity';

export interface Pattern<T> extends DurationEntity {
  timeline?: Timeline<T>
  innerPatternTimeline?: Timeline<Pattern<T>>
}

export type PatternMap<T> = {
  [P in keyof T]: Pattern<T[P]>;
};
